rex/resource.py

```diff
@@ -148,9 +212,95 @@ def _check_slice(ds_slice):
 
         return ds_slice, ds_idx
 
+    def _extract_list_slice(self, ds_slice):
+        """
+        Optimize and extract list slice request along a single dimension
+
+        Parameters
+        ----------
+        ds_slice : int | slice | list | ndarray | tuple
+            What to extract from ds, each arg is for a sequential axis
+
+        Returns
+        -------
+        out : ndarray
+            Extracted array of data from ds
+        """
+        out_slices = []
+        chunks = self.chunks
+        sort_idx = []
+        list_len = None
+        if chunks:
+            for i, ax_slice in enumerate(ds_slice):
+                c = chunks[i]
+                if isinstance(ax_slice, (list, np.ndarray)):
+                    if not isinstance(ax_slice, np.ndarray):
+                        ax_slice = np.array(ax_slice)
+
+                    idx = np.argsort(ax_slice)
+                    sort_idx.append(np.argsort(idx))
+                    ax_slice = ax_slice[idx]
+                    diff = np.diff(ax_slice) > c
+                    if np.any(diff):
+                        pos = np.where(diff)[0] + 1
+                        ax_slice = np.split(ax_slice, pos)
+                        list_len = len(ax_slice)
+                elif isinstance(ax_slice, slice):
+                    sort_idx.append(slice(None))
+
+                out_slices.append(ax_slice)
+        else:
+            out_slices = ds_slice
+
+        if list_len is not None:
+            out_slices = self._make_list_slices(out_slices, list_len)
+            axis = [i for i, s in enumerate(sort_idx)
+                    if not isinstance(s, slice)][0]
+            out = None
+            for s in zip(*out_slices):
+                arr = self._extract_ds_slice(s)
+                if out is None:
+                    out = arr
+                else:
+                    out = np.append(out, arr, axis=axis)
```

Comment:
It would be much faster to init the full size np array as zeros or zeros *= np.nan than to append a new array every time (np.append 1 of 2)